### ISSUE TITLE
Rule proposal: [prefer-at] Rule to prefer array.at(-1) over array[array.length - 1]

### ISSUE THREAD

**Issue by JoshuaKGoldberg** [typescript-eslint/typescript-eslint#6401]
### Before You File a Proposal Please Confirm You Have Done The Following...

- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22enhancement%3A+new+plugin+rule%22) and found none that match my proposal.
- [X] I have searched the [current rule list](https://typescript-eslint.io/rules/#supported-rules) and found no rules that match my proposal.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### My proposal is suitable for this project

- [X] My proposal specifically checks TypeScript syntax, or it proposes a check that requires type information to be accurate.
- [X] My proposal is not a "formatting rule"; meaning it does not just enforce how code is formatted (whitespace, brace placement, etc).
- [X] I believe my proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal).

### Description

#6211 shows a use of `array.at(-1)`, as a cleaner equivalent to `array[array.length - 1]`. https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/at

Proposal: let's add a typed lint rule that, when it sees something like `array[array.length - 1]` on an Array-typed value, proposes fixing to `array.at(-1)`?

### Fail Cases

```typescript
declare const array: string[];

array[array.length - 1];
```

### Pass Cases

```typescript
declare const array: string[];

array.at(-1);

// [] indices are reasonable
array[0];

// Don't look through functions
const getArray = () => array;
getArray()[getArray().length - 1];
```

### Additional Info

Reference for an existing similar rule: https://typescript-eslint.io/rules/prefer-includes.

I don't think we should enable this in _recommended_ or _strict_ configs until we only support Node versions that include `.at`. Per #6396 this wouldn't make sense for users who compile to Node 14 and don't include a polyfill.

**Comment by sviat9440** [comment]
> Why ?  (sorry )

I think we can't guarantee that the function always returns the same array. But it seems to be a rarity. In most cases, it can be replaced with .at(-1). Let's see what others have to say.



> Also, do we want to limit to `-1`?

Hmm, indeed, if `arr[arr.length - 1]` is equivalent to `array.at (-1)`, then why not do the same for other numbers?
@JoshuaKGoldberg , what do you think?

**Comment by JoshuaKGoldberg** [comment]
> Why 🤔 (sorry 😅)

> I think we can't guarantee that the function always returns the same array. But it seems to be a rarity. In most cases, it can be replaced with .at(-1). Let's see what others have to say.

Haha, yeah, these very simplified code snippets often look ridiculous. But we have to imagine that _sometimes_ there will be _some_ code _somewhere_ that implements those odd patterns. And if that code is just barely not bizarre enough to be ignored, then we have to consider it as a valid edge case for the rule.

In this case, it's not uncommon for people to come from traditional OOP style programming paradigms and write code like:

```ts
declare class Box<T> {
  constructor(private readonly value: T);
  getValue(): T;
  // maybe some other methods, like updateValue(newValue: T)
}

const box = new Box([Math.random(), Math.random()]);
const lastRandomNumber = box.getValue()[box.getValue().length - 1];
```

...but we as the linter might not know that `getValue()` always returns the same thing.

> Hmm, indeed, if `arr[arr.length - 1]` is equivalent to `array.at(-1)`, then why not do the same for other numbers? @JoshuaKGoldberg , what do you think?

Yeah makes sense to me! Let's say any constant literal or number after the ` - `, such as `1`, `2`, `i`, etc.

**Comment by sviat9440** [comment]
> but we as the linter might not know that `getValue()` always returns the same thing.

How about creating an option: ignore functions or not?

**Comment by Josh-Cena** [comment]
> ignore functions

I don't think so. It's really hard to generalize whether a codebase has all side-effect-free functions or not, and configuring such an option could easily break code in the future. We should just assume all functions are side-effect-ful, because I don't feel like sane code that's meant to be side-effect-free would be written that way anyway.

**Comment by sviat9440** [comment]
> that's meant to be side-effect-free

We can just display a warning for functions without autofocus.

**Comment by Josh-Cena** [comment]
Sounds good to me. We can autofix "definitely safe" things (which honestly only includes identifiers, but we can assume member expressions are generally safe as well), while keeping others as suggestions.

**Comment by bradzacher** [comment]
Semantically I think that this rule should be safe to fix through a function call.

There really no reasonable code which should look like `fn()[fn().length - 1]`.
Thinking through it - the only time that wouldn't be safe to fix is if the function returns some predictable pattern of arrays. We're talking about an edge of an edge case here, IMO.

If you wanted to split hairs about safety:
```ts
const x = {
  get arr() {
    return Array(Math.random());
  },
};

x.arr[x.arr.length - 1];
```
Technically it's not safe to fix member expressions due to getters or to proxies!

I think we can safely assume that if you call the same function in the indexer and the expression then you're working with the same value.

Definitely agree we can add an option to disable that behaviour - but I would expect it to be an opt-out, I.e. Default on, because it's likely the most common case.

**Comment by Josh-Cena** [comment]
> If you wanted to split hairs about safety

That's exactly what I said with "we can assume member expressions are generally safe" :) But IMO we have to make trade-offs here about what's likely and what's not.

**Comment by omril1** [comment]
1. I think the rule should apply to strings as well
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/at
2. There is a similar rule in https://github.com/sindresorhus/eslint-plugin-unicorn/blob/main/docs/rules/prefer-at.md

**Comment by sviat9440** [comment]
How about this case?

```
      declare let a: string;

      const lastItem = (a += 'test')[(a += 'test').length - 1];
```

Should it be converted to a: ?
```
      declare let a: string;

      const lastItem = (a += 'test').at(-1);
```

Now - it's being transformed

**Comment by bradzacher** [comment]
@sviat9440 Is that a hypothetical case, or code you've legitimately encountered in the wild?

**Comment by sviat9440** [comment]
It is hypothetical case

**Comment by bradzacher** [comment]
This code is also correct code that is unsafe to fix:
```ts
let x = 0;
const arr = [1,2,3];

const lastItem1 = (x += 1, arr)[(x += 1, arr).length - 1];

// or

declare let cond: boolean;
const lastItem2 = (cond ? (cond = !cond, []) : (cond = !cond, arr))[
	(cond ? (cond = !cond, []) : (cond = !cond, arr)).length - 1
];
```

My point is that we could sit here all day dreaming up "semantically correct code a user may write that we shouldn't fix"! The joy of an expression location is that there is literally an infinite number of permutations of code that has side-effects and is semantically correct in the location!

My take on this is that we're talking about code that people won't ever write in the real world - so it's not worth the maintenance burden and complexity of handling and testing for them.

If someone comes to us with a real world example of their code that they have to write in such an insane style - then, and only then we could consider handling the case (though TBH I'd probably tell them to just use a disable comment...). Otherwise I think we should just "assume good intent" and assume that in the case of `T[T - i]` - `T` is an expression without side-effects which I would guess is an assumption that covers 99.99999999% of real code.

**Comment by ronami** [comment]
What do you think about enforcing the use of `.at()` for accessing array elements over angle brackets in general?

A very similar rule already exists on [`eslint-plugin-unicorn`](https://github.com/sindresorhus/eslint-plugin-unicorn/blob/main/docs/rules/prefer-at.md) (requires enabling `checkAllIndexAccess`, as the rule only checks negative indexes by default), but since it doesn't have access to the type system, it misses some edge cases:

```ts
declare const s: string;
declare const a: string[];
declare const n: number;

// inability to differentiate between string and array access
s[1];

// inability to fix dynamic indexes (other than the special case `array[array.length - i]`)
a[n];

// wrongly auto-fixes odd cases
const o = { 1: 1, 2: 2 };
o[1];
```

Additionally, given access to the type system, the rule can potentially auto-fix places where the contextual type is `T | undefined` (similar to the suggestion in #8979), along with auto-fixing everything if `noUncheckedIndexedAccess` is enabled (essentially makes this a stylistic change). Other cases can be a suggestion.

A rule like this can be a convenient way to gradually opt into the more strict `noUncheckedIndexedAccess` flag. Another motivation is to make the code base more consistent as array index access is standardized.

**Comment by JoshuaKGoldberg** [comment]
That seems pretty strong to me. `noUncheckedIndexedAccess` is IME really, _really_ annoying to enable in most apps. It's not something most people would want. I don't think it would even fit in our `strict*` configs.

Maybe that could be a followup issue to add an opt-in option for the rule? We could name the rule something broad like `prefer-array-at` to leave the door open for an option like `indices: '-1' | 'negative' | 'always"`?

**Comment by Josh-Cena** [comment]
I can understand preferring `array.at(-i)` over `array[array.length - i]`, but I don't know why we would be separating `"-1"` from `"negative"` in general. I personally would strongly advise against `"always"` because of optimization and readability, but I would not be against that as an option.
